src/Proto.Cluster/Gossip/Consensus.cs

```diff
@@ -0,0 +1,78 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="ClusterConsensus.cs" company="Asynkron AB">
+//      Copyright (C) 2015-2021 Asynkron AB All rights reserved
+// </copyright>
+// -----------------------------------------------------------------------
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Proto.Cluster.Gossip
+{
+    public interface IConsensusHandle<T> : IDisposable
+    {
+        Task<(bool consensus, T value)> TryGetConsensus(CancellationToken ct);
+        Task<(bool consensus, T value)> TryGetConsensus(TimeSpan maxWait, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Clear any current consensus to prevent stale reads
+        /// </summary>
+        public void TryResetConsensus();
+    }
+
+    internal class GossipConsensusHandle<T> : IConsensusHandle<T>
+    {
+        private TaskCompletionSource<T> _consensus = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        private readonly Action _deregister;
+
+        public GossipConsensusHandle(Action deregister) => _deregister = deregister;
+
+        internal void TrySetConsensus(object consensus)
+        {
+            if (_consensus.Task.IsCompleted && _consensus.Task.Result?.Equals(consensus) != true)
+            {
+                TryResetConsensus();
+            }
+
+            //if not set, set it, if already set, keep it set
+            _consensus.TrySetResult((T) consensus);
+        }
+
+        public void TryResetConsensus()
+        {
+            //only replace if the task is completed
+            var current = _consensus;
+
+            if (current.Task.IsCompleted)
+            {
+                var taskCompletionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+                Interlocked.CompareExchange(ref _consensus, taskCompletionSource, current);
+                
+            }
+        }
+
+        public async Task<(bool consensus, T value)> TryGetConsensus(CancellationToken ct)
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                var t = _consensus.Task;
+                // ReSharper disable once MethodSupportsCancellation
+                await Task.WhenAny(t, Task.Delay(500));
+                if (t.IsCompleted)
+                    return (true, t.Result);
+            }
+
+            return (false, default);
+        }
+        
+        public async Task<(bool consensus, T value)> TryGetConsensus(TimeSpan maxWait, CancellationToken cancellationToken)
+        {
+            var t = _consensus.Task;
+            await Task.WhenAny(t, Task.Delay(maxWait, cancellationToken)).ConfigureAwait(false);
```

Comment:
We should probably introduce some abstraction for these WhenAny + Delay.
Because if the first task `t` is already completed, there will be delay tasks spawned that doesn't get cancelled.
_potentially_ in some bussy section of a system which spams TryGetConsensus, this could lead to a lot of tasks floating around.

Not really caused by this PR as we have had this a bit of everywhere.

Reply:
We do have `Proto.Utils.TaskExtensions.WithTimeout`, but it does not quite solve the same problem. It also throws TimeoutException which is not what we want here.

How would the ideal abstraction look for something like this? We want to fall back to a default value if maxwait expires. But we also want to be able to cancel the operation if it is no longer relevant..